### PR TITLE
Build serenity Toolchain, Kernel, Userland and Root FS natively on macOS

### PR DESCRIPTION
**Pull request by Pagghiu**
This pull request enables building serenity toolchain, kernel, userland and rootfs on macOS.
Some background discussion can be found visiting #808.

This involves using clang as bootstrap compiler to build GCC and the IPCCompiler tools, build everything and then mount and create the rootfs.

Notes:
- I have explicitly downloaded GCC dependencies to static link them, building a gcc cross-compiler that doesn't depend on any dylib installed. I've had problems in the past with brew installing different and incompatible versions of the support libraries (isl, mpfr etc.) so that's my suggested way of building GCC on macOS. Also the Linux build could statically link them to, to mitigate similar issues, if they exist.
- Patches in the build-root-filesystem.sh shell scripts are needed because macOS  mknod doesn't support -m option. 
- Some gnu tools and ext2 FUSE drivers need also to be installed, to allow building the root fs image, and that's what BuildFuseExt2.sh is for.

Everything should be compiling and working fine on macOS 10.14.6 and XCode 11.0 (my computer)
Some people have been able to build and run serenity from scratch on macOS using this pull request, so it's hopefully a little less "it works on my computer-esque" than usual ;)
I've been adding build instructions in the Readme file (not being sure if this is acceptable), but they can be moved anywhere it makes sense for them to be.

Finally, I've tried to be very conservative with all the changes, using a good amount of #ifdef and such.
I have been testing the build on a Linux Virtual Machine and it looks like I have not been braking anything. Nevertheless I suggest to run a full rebuild on a real Linux machine before merging it ;)
 
**Known Issues**
- AK Tests have been disabled because clang fails compiling them

qemu on macOS runs pretty slowly because:
- qemu kvm accelerator is not available on macOS
- qemu hypervisor framework accelerator doesn't work when using 32 bit virtualization 
- qemu intel haxm backend crashes serenity during kernel boot (to be investigated...)

### CODE REVIEW

**Comment by remyabel** [comment]
Drive-by comment:

> I have explicitly downloaded GCC dependencies to static link them, building a gcc cross-compiler that doesn't depend on any dylib installed. I've had problems in the past with brew installing different and incompatible versions of the support libraries (isl, mpfr etc.) so that's my suggested way of building GCC on macOS.

Have you tried using [`contrib/download_prerequisites`](https://github.com/gcc-mirror/gcc/blob/master/contrib/download_prerequisites)? It downloads the prerequisites, builds and links them in-tree.

**Comment by Pagghiu** [comment]
Yes, I am basically doing the same as the "download_prerequisites" script.
The versions I am manually specifying in the BuildIt.sh are the same as in download_prerequisites.
I think back in the days when I did my "how to build GCC the way I like" I did it this way just to have control on where tarballs are actually being downloaded.

I can try testing that download_prerequisites script instead, so that the linked library versions will be "in sync" with the correct GCC version even when the toolchain will be upgraded to a later GCC version.

Thanks for the input!

File: Toolchain/BuildFuseExt2.sh [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+die() {
+    echo "die: $*"
+    exit 1
+}
+
+if [[ "$OSTYPE" != "darwin"* ]]; then
+    die "This script makes sense to be run only on macOS"
+fi
+export PATH=/opt/gnu/bin:$PATH
```
Comment (bugaevc) [inline Toolchain/BuildFuseExt2.sh]: I think it's an established convention that building Serenity does not impact your system, i.e.all the scripts don't touch anything outside of the Serenity root.

File: ReadMe.md [hunk 2]
```diff
@@ -77,6 +78,24 @@ sudo apt-get install gcc-8 g++-8
...
+brew cask install osxfuse
+Toolchain/BuildFuseExt2.sh
+```
+
+Notes: 
+- Assuming xcode and xcode-tools are installed (git is required by some scripts)
+- coreutils is needed for building gcc
+- qemu is needed to run the actual OS. As an alternative you can Build it using the BuildQemu.sh script
+- osxfuse and BuildFuseExt2.sh are needed if you want to build the root filesystem disk image natively on macOS. This allows mounting an EXT2 fs and also installs commands like mke2fs that are not available on stock macos. 
+- If you install some commercial EXT2 macOS fs handler instead of osxfuse and fuse-ext2, you will need to install e2fsprogrs using brew to obtain mke2fs.
```
Comment (bugaevc) [inline ReadMe.md]: Please format script names and such with `backticks` as it's done in other sections

File: Libraries/LibCore/CLocalServer.cpp [hunk 3]
```diff
@@ -71,7 +74,19 @@ bool CLocalServer::listen(const String& address)
 
     int rc;
 
+#ifdef SOCK_NONBLOCK
     m_fd = socket(AF_LOCAL, SOCK_STREAM | SOCK_NONBLOCK | SOCK_CLOEXEC, 0);
+#else
+    m_fd = socket(AF_LOCAL, SOCK_STREAM, 0);
+#ifdef _WIN32
```
Comment (bugaevc) [inline Libraries/LibCore/CLocalServer.cpp]: Is there a Windows port in the works too? 😱

File: Libraries/LibCore/CIODevice.h [hunk 4]
```diff
@@ -5,7 +5,7 @@
 #include <LibCore/CObject.h>
 
 class CIODevice : public CObject {
-    C_OBJECT(CIODevice)
+    C_OBJECT_ABSTRACT(CIODevice)
```
Comment (bugaevc) [inline Libraries/LibCore/CIODevice.h]: Why is this needed?

Comment (bugaevc) [reply Libraries/LibCore/CIODevice.h]: Must be because of some warning that clang emits about trying to construct this in `construct()`, even though the template never gets instantiated

File: Toolchain/BuildFuseExt2.sh (hunk 1 above)
Comment (Pagghiu) [reply Toolchain/BuildFuseExt2.sh]: Well, I am a big fan of self contained systems and softwares,  but on macOS the quickest way I've found to create and mount an ext2 volume is to install fuse-ext2, that installs as a "system" FUSE driver/backend.
I think you can see this as a "prerequisite", exactly like the apt-get for libraries and GCC8 that are listed on ReadMe in the Linux session.

Probably we can redirect most of these scripts to build and install things locally, but I am not sure if one can get entire fuse-ext2 installed without putting its libraries into a "system" folder.
If this behavior is a problem we can put just a stub here with a message linking to the script living in some other repo.

Something that would be cool to do is to be able compiling and using use the EXT2 code outside of Kernel, to build the image from scratch with an internal helper tool without external dependencies. I am not sure how difficult would be to do it, I would need to look at the code.
That would make it easily possible to build and run Serenity on Windows too!

File: Libraries/LibCore/CLocalServer.cpp (hunk 3 above)
Comment (Pagghiu) [reply Libraries/LibCore/CLocalServer.cpp]: Why not? 😄

File: Libraries/LibCore/CIODevice.h (hunk 4 above)
Comment (Pagghiu) [reply Libraries/LibCore/CIODevice.h]: Exactly, clang complains about that construct() function on an abstract class.

File: ReadMe.md (hunk 2 above)
Comment (Pagghiu) [reply ReadMe.md]: Ok, sure

File: Toolchain/BuildFuseExt2.sh (hunk 1 above)
Comment (bugaevc) [reply Toolchain/BuildFuseExt2.sh]: > I think you can see this as a "prerequisite", exactly like the apt-get for libraries and GCC8 that are listed on ReadMe in the Linux session.

Sure; but the difference is that apt is already *the* right way to install packages (on Debian variants), it does install them to the right directories and uses the right settings / build options.

You're building something yourself, and you're hardcoding stuff like paths; I may want the software in say `/usr/local` instead of `/opt/gnu`. I pick a Linux distro based on how I like these things organized, then I install gcc dependencies in accordance with my distro's policies; but this script basically hardcodes its own policy, which would be okay if it was Serenity-specific and wouldn't touch system-wide stuff. Maybe I do already have some of the necessary software installed, after all, what would happen then?

So I would like this script to at least asks the user for their preferred PATH/prefix, maybe as an optional argument to the script, maybe as an interactive thing you have to confirm.

Besides, isn't this stuff (automake, autoconf, libtool, ext2fuse, ext2fsprogs) already available in Homebrew?

File: Libraries/LibCore/CLocalServer.cpp (hunk 3 above)
Comment (bugaevc) [reply Libraries/LibCore/CLocalServer.cpp]: Do we even need `CLocalServer` on the host for something?

File: Libraries/LibCore/CIODevice.h (hunk 4 above)
Comment (bugaevc) [reply Libraries/LibCore/CIODevice.h]: Well please do that, and other clang-related changes, in a separate commit with a descriptive commit message :)

**Comment by bugaevc** [comment]
If my inline comments sound overly negative, let me just say that this is very cool 😀

File: Libraries/LibCore/CLocalServer.cpp (hunk 3 above)
Comment (Pagghiu) [reply Libraries/LibCore/CLocalServer.cpp]: Not directly, but it's part of the base libraries built for host tools (IPCCompiler etc.).
I think it's easier to fix compiling that file on non-Linux hosts rather than adding complexity to makefiles when compiling host tools.

Comment (bugaevc) [reply Libraries/LibCore/CLocalServer.cpp]: I anticipate that Unix-specific parts of LibCore (servers, sockets, pipes and stuff like that — not utilities like `CObject` and `CFile`) are only going to become more Serenity-specific; and there should not be any need for them during code generation on the host, so I would opt for only building them for Serenity.

But then maybe they're *supposed* to be ported as a part of Lagom, so let's defer to @awesomekling's judgement :)

Comment (awesomekling) [reply Libraries/LibCore/CLocalServer.cpp]: Hmm. Please fix LibCore mac stuff in a separate commit if it’s not too invasive. I agree it’s easier to just fix the build instead of fixing IPCCompiler and pals to not use LibCore. :)

**Comment by Pagghiu** [comment]
> If my inline comments sound overly negative, let me just say that this is very cool 😀

No problem absolutely! 😃 

I am clearly not aware of all the good pull-request-etiquette that is acceptable for this project, and I am learning from everybody's comments so that I can target it more accurately.

**Comment by bugaevc** [comment]
Here's another tip then: please reformat the code changes with clang-format. In particular the preprocessor directives should be aligned like this:
```cpp
    code code code
#ifdef FOO_BAR
    conditional code
#endif
    more code
```

File: Toolchain/BuildFuseExt2.sh (hunk 1 above)
Comment (Pagghiu) [reply Toolchain/BuildFuseExt2.sh]: I have tried installing as much as possible from brew.
Unfortunately `fuse-ext2` brew formula is not available, and it will still need the custom install script.

File: ReadMe.md (hunk 2 above)
Comment (Pagghiu) [reply ReadMe.md]: I have tried formatting it using `backticks`, please let me know if it looks more or less right now.

File: Libraries/LibCore/CLocalServer.cpp (hunk 3 above)
Comment (Pagghiu) [reply Libraries/LibCore/CLocalServer.cpp]: I have tried splitting the fixes in multiple specific commits, as per your requests.

File: Libraries/LibCore/CIODevice.h (hunk 4 above)
Comment (Pagghiu) [reply Libraries/LibCore/CIODevice.h]: Done!

**Comment by Pagghiu** [comment]
I have tried following all suggestions, just let me know if there's anything else that is needed.

**Comment by awesomekling** [comment]
Okay let's merge this! If someone has more suggestions we can follow up with more fixes :)
